capsules/src/usb/ctap.rs

```diff
@@ -0,0 +1,391 @@
+                transfer_type: TransferType::Interrupt,
+                max_packet_size: 64,
+                interval: 5,
+            },
+            EndpointDescriptor {
+                endpoint_address: EndpointAddress::new_const(0x81, TransferDirection::HostToDevice),
+                transfer_type: TransferType::Interrupt,
+                max_packet_size: 64,
+                interval: 5,
+            },
+        ]];
+
+        let (device_descriptor_buffer, other_descriptor_buffer) =
+            descriptors::create_descriptor_buffers(
+                descriptors::DeviceDescriptor {
+                    vendor_id: vendor_id,
+                    product_id: product_id,
+                    manufacturer_string: 1,
+                    product_string: 2,
+                    serial_number_string: 3,
+                    class: 0x03, // Class: HID
+                    max_packet_size_ep0: MAX_CTRL_PACKET_SIZE,
+                    ..descriptors::DeviceDescriptor::default()
+                },
+                descriptors::ConfigurationDescriptor {
+                    ..descriptors::ConfigurationDescriptor::default()
+                },
+                interfaces,
+                endpoints,
+                Some(&HID_DESCRIPTOR),
+                None,
+            );
+
+        CtapHid {
+            client_ctrl: ClientCtrl::new(
+                controller,
+                device_descriptor_buffer,
+                other_descriptor_buffer,
+                Some(&HID_DESCRIPTOR),
+                Some(&REPORT),
+                LANGUAGES,
+                strings,
+            ),
+            buffers: [Buffer64::default(), Buffer64::default()],
+            tx_buffer: TakeCell::empty(),
+            tx_len: Cell::new(0),
+            tx_offset: Cell::new(0),
+            tx_client: OptionalCell::empty(),
+            rx_buffer: TakeCell::empty(),
+            rx_len: Cell::new(0),
+            rx_offset: Cell::new(0),
+            rx_client: OptionalCell::empty(),
+        }
+    }
+
+    #[inline]
+    fn controller(&self) -> &'a U {
+        self.client_ctrl.controller()
+    }
+
+    #[inline]
+    fn buffer(&'a self, i: usize) -> &'a [VolatileCell<u8>; 64] {
+        &self.buffers[i - 1].buf
+    }
+}
+
+impl<'a, U: hil::usb::UsbController<'a>> hil::usb::Client<'a> for CtapHid<'a, U> {
+    fn enable(&'a self) {
+        // Set up the default control endpoint
+        self.client_ctrl.enable();
+
+        // Setup buffers for IN and OUT data transfer.
+        self.controller()
+            .endpoint_set_in_buffer(ENDPOINT_IN_NUM, self.buffer(ENDPOINT_IN_NUM));
+        self.controller()
+            .endpoint_in_enable(TransferType::Bulk, ENDPOINT_IN_NUM);
+
+        self.controller()
+            .endpoint_set_out_buffer(ENDPOINT_OUT_NUM, self.buffer(ENDPOINT_OUT_NUM));
+        self.controller()
+            .endpoint_out_enable(TransferType::Bulk, ENDPOINT_OUT_NUM);
```

Comment:
This section is misleading, because in reality `ENDPOINT_IN_NUM == ENDPOINT_OUT_NUM`. In that case, the correct way to call the HIL is to use `endpoint_in_out_enable`.

Also, at least for the OpenSK case, this should be an Interrupt transfer, not Bulk.

That's the kind of things I'm worried about in terms of the lack of robustness of this pull request.

Reply:
Fixed.